assembly/src/mast_forest_builder.rs

```diff
@@ -346,7 +346,13 @@ impl MastForestBuilder {
                     decorators.push((op_idx + operations.len(), decorator));
                 }
                 for batch in basic_block_node.op_batches() {
-                    operations.extend_from_slice(batch.ops());
```

Comment:
- Continuation of the conversation in the "main PR thread" so that we get a single thread for it -

> More generally, we must ensure that any MASM instruction be represented in the textual format so that the AST can be round-tripped (i.e. we should not emit MASM containing e.g. nop when the user can't write the same MASM by hand).

If we agreed that there was no reason for a user to *want* to write a `nop` (are there?), and that the only 2 ways to get a `Operation::Noop` in a `BasicBlockNode` is if it was "inserted by the processor" (for the 2 reasons discussed in this PR's issue), then we could remove `Instruction::Nop` from MASM altogether. This would simplify how we handle block merging, etc.

Reply:
The compiler currently uses `nop` to attach trace events for capturing call frame information. I haven't explored whether that can be done with equivalent fidelity using other approaches, but AFAIK we don't yet have a way to attach a trace event to the _end_ of a block (only entering it), which is we use use `nop` for that purpose currently.

I agree though that in principle, we could remove `nop` from MASM if there is no reason to ever write it by hand - but are we sure there aren't valid reasons to do so? In traditional assemblers, it's usually to pad out code to a certain alignment, or to leave space for hot patching - I'm not sure what the equivalent in Miden would be, maybe padding out a procedure so it hits a certain number of cycles or something? 🤷

Reply:
> but are we sure there aren't valid reasons to do so? In traditional assemblers, it's usually to pad out code to a certain alignment, or to leave space for hot patching - I'm not sure what the equivalent in Miden would be, maybe padding out a procedure so it hits a certain number of cycles or something?

Yeah *perhaps* in some cases someone might want to use it to differentiate the MAST roots of 2 otherwise identical procedures? Might be useful to workaround some VM issues or something. I guess it's probably safer to keep it in that case 😔